### ISSUE TITLE
TST doesn't detect unload by "suspend tabs" addons (which support TST API)

### ISSUE THREAD

**Issue by perdolka** [piroor/treestyletab#1903]
## Short description
TST doesn't always detect unload by "suspend tabs" addons (which support TST API)
Thus, tab is actually unloaded however, the [marking of unloaded tabs with gray shading](https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules#change-styling-of-pending-unloaded-tabs-1363) doesn't always works. It happens randomly.

## Steps to reproduce
 1. Start Firefox with clean profile.
 2. a) Install TST.
b) Open TST options, use [Code snippet](https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules#change-styling-of-pending-unloaded-tabs-1363),
for example, this:
```
.tab.discarded 
{
  opacity: 0.2;
}
```
 3. a) [Install Auto Tab Discard](https://addons.mozilla.org/firefox/addon/auto-tab-discard/versions/) 0.2.8
b) Open ATD settings. set automatic discarding time to 10 seconds. Just for speed up the experiment.
 4. a) Open google.com, 
b) type some search query, like "test" 
c) open ~10 links from search results
d) don't visit them, wait for automatic discarding. 

Example screenshot:
<a href="http://piccy.info/view3/12354374/1c83b750134cffee1a492ac627dd60e8/" target="_blank"><img src="http://i.piccy.info/i9/37419df08bd8e77888668f71bcb6357e/1527077992/11374/1246306/sshot_124_min.png" alt="Piccy.info - Free Image Hosting" border="0" /></a><a href="http://i.piccy.info/a3c/2018-05-23-12-19/i9-12354374/538x210-r" target="_blank"><img src="http://i.piccy.info/a3/2018-05-23-12-19/i9-12354374/538x210-r/i.gif" alt="" border="0" /></a>

## Some references:
https://github.com/rNeomy/auto-tab-discard/issues/26
https://github.com/Lej77/tab-unloader-for-tree-style-tab/issues/3

## Environment
 * Platform (OS): Windows 10 (x64)
 * Version of Firefox: 60.0.1 (x64)
 * Version (or revision) of Tree Style Tab: 2.4.21
 * Version of  Auto Tab Discard: 0.2.8

**Comment by piroor** [comment]
I tried the steps described, but I couldn't reproduce the problem. Tabs opened from search results are unloaded after some delay and style rules for `.tab.discarded`. CSS codes I actually used are:

```css
.tab.discarded { background: green !important; }
.tab.discarded * { color: white !important; }
```

**Comment by perdolka** [comment]
This problem doesn't always occur. 
I can't reproduce the "issue algorithm" to get it always.

Temporary workaround - the dimming works correctly if you call the discard method once again,
manually, on the unloaded BUT not dimmed tab.

Maybe @Lej77 thoughts can help:
> The issue is probably still that Firefox tells extensions that the tab is loaded even though it isn't. This seems to happen if a content script is sending a message when it is being unloaded and my pull request prevents one such message. 
> If this is the reason for this issue then it would be quite hard to fix with in Tree Style Tab's code. If the problem is because of Firefox's tracking then I see two solutions. Either stop the message that is being sent when the tab is unloaded or call the discard function twice (once to unload the tab and another to fix the incorrect Firefox state).

**Comment by Lej77** [comment]
@perdolka I'm pretty sure this is bug of Firefox itself. It makes it so that tabs have discarded set to false even though they aren't loaded. I just released a new version of [Tab Unloader for Tree Style Tab](https://addons.mozilla.org/firefox/addon/tab-unload-for-tree-style-tab/) that includes a workaround that should fix such an issue for all unloaded tabs (even those not unloaded by my extension). It does this by listening to the `browser.tabs.onUpdate` event and if a tab has `discarded` set to `true` and then immediately after have `discarded` set to `false` and no `tab.status` or `tab.url` changes are observed after it is "loaded" then its probably this bug so it then unloads the tab again which sets `discarded` to `true`.

**Comment by piroor** [comment]
@Lej77 Hmm, is the bug tracked on bugzilla.mozilla.org?

**Comment by Lej77** [comment]
@piroor I haven't found it there but I haven't looked that much either. Anyway I wrote a small test extension to reproduce the false loaded state.

`manifest.json`:
```json
{
  "manifest_version": 2,
  "name": "Mark tab as loaded when unloaded",
  "version": "1.0",
  "applications": {
    "gecko": {
      "strict_min_version": "58.0"
    }
  },
  "permissions": [
    "menus",
    "<all_urls>"
  ],
  "background": {
    "scripts": [
      "background.js"
    ]
  }
}
```

`background.js`:
```javascript
browser.menus.create({
  id: "fake-unload-tab",
  title: "Unload and mark as loaded",
  contexts: ["tab"]
});

browser.menus.onClicked.addListener(async function (info, tab) {
  switch (info.menuItemId) {
    case 'fake-unload-tab': {
      browser.tabs.discard(tab.id);
      browser.tabs.executeScript(tab.id, { code: `browser.runtime.sendMessage({});`, runAt: 'document_start', });
    } break;
  }
});

browser.runtime.onMessage.addListener((message, sender) => {

});
```

Only tabs with pages that can have content scripts can be put in the fake loaded state since a message needs to be sent from the page when it is being unloaded. Also the problem can't be reproduced 100% of the time but it appears most of the time. Sometimes it doesn't work for a particular tab or page so test different ones if it doesn't work. (Note: the context menu item won't work for the active tab since it will be ignored by the `browser.tabs.discard` function.)

**Comment by piroor** [comment]
1464992 - Immediate operations after tabs.discard() disposes tab's "discarded" attribute but the tab is actually still discarded
https://bugzilla.mozilla.org/show_bug.cgi?id=1464992
I've reported this problem to bugzilla.mozilla.org.

**Comment by irvinm** [comment]
@piroor I think we could probably close this one.
